Enable XML_ATTR_INFO in build

MohammedKhajapasha opened the pull request:
enable XML_ATTR_INFO in build to cover XML_ATTR_INFO in CI

Signed-off-by: Mohammed Khajapasha <mohammed.khajapasha@intel.com>

hartwork commented on expat/coverage.sh:
```diff
@@ -38,8 +38,10 @@ _get_coverage_dir() {
+
     ${unicode_enabled} \
-            && configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )
+            && configure_args+=( CPPFLAGS+='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )
```
That second `+=` looks wrong to me. Does configure support that?

Also gave it a try:
```console
# ./configure CPPFLAGS=X CPPFLAGS+=Y
configure: error: invalid variable name: `CPPFLAGS+'
```

MohammedKhajapasha replied:
Updated review comments and uploaded the change

hartwork commented on expat/configure.ac:
```diff
@@ -221,6 +221,8 @@ AC_DEFINE([XML_DTD], 1,
 AC_DEFINE([XML_DEV_URANDOM], 1,
           [Define to include code reading entropy from `/dev/urandom'.])
+AC_DEFINE([XML_ATTR_INFO], 1,
+          [Define to allow retrieving the byte offsets for attribute names and values.])
```
This code force-enables `XML_ATTR_INFO` when compiling with GNU Autotools. The CMake version above has a switch, in contrast. It would be nice to have a switch on both sides.

MohammedKhajapasha replied:
Update review comments in the change

MohammedKhajapasha commented:
Addressed below comments and updated #244 pull request.
Please review the same.
--
Regards
Khajapasha
From: Sebastian Pipping [mailto:notifications@github.com]
Sent: Wednesday, June 19, 2019 9:17 PM
To: libexpat/libexpat <libexpat@noreply.github.com>
Cc: Khajapasha, Mohammed <mohammed.khajapasha@intel.com>; Author <author@noreply.github.com>
Subject: Re: [libexpat/libexpat] Enable XML_ATTR_INFO in build (#264)


@hartwork requested changes on this pull request.

I notice that I'm not 100% sure about these open questions:

  *   should the new default be ON or OFF for XML_ATTR_INFO? It's been OFF so far but extending the static .def files would need ON for a default (unless we drop the Visual Studio files build system and only generate those through CMake)
  *   how critical is it to get both XML_ATTR_INFO covered by CI (#244<https://github.com/libexpat/libexpat/issues/244>)? I'm tending towards need for coverage. The pull request does not yet do that.
  *   should we drop and auto-enable XML_ATTR_INFO code for everyone, instead?

I appreciate your input!

CC @RMJ10<https://github.com/RMJ10> @kwaclaw<https://github.com/kwaclaw>

________________________________

In expat/coverage.sh<https://github.com/libexpat/libexpat/pull/264#discussion_r295361288>:

>      ${unicode_enabled} \

-            && configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )

+            && configure_args+=( CPPFLAGS+='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )

That second += looks wrong to me. Does configure support that?

Also gave it a try:

# ./configure CPPFLAGS=X CPPFLAGS+=Y

configure: error: invalid variable name: `CPPFLAGS+'

________________________________

In expat/configure.ac<https://github.com/libexpat/libexpat/pull/264#discussion_r295368766>:

> @@ -221,6 +221,8 @@ AC_DEFINE([XML_DTD], 1,

           [Define to make parameter entity parsing functionality available.])

 AC_DEFINE([XML_DEV_URANDOM], 1,

           [Define to include code reading entropy from `/dev/urandom'.])

+AC_DEFINE([XML_ATTR_INFO], 1,

+          [Define to allow retrieving the byte offsets for attribute names and values.])

This code force-enables XML_ATTR_INFO when compiling with GNU Autotools. The CMake version above has a switch, in contrast. It would be nice to have a switch on both sides.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/libexpat/libexpat/pull/264?email_source=notifications&email_token=AMAHZCPZTRO5GJ3M7MB6TI3P3JIGXA5CNFSM4HZH6WT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4A6ILI#pullrequestreview-251782189>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMAHZCJKLKRLQ22JWM45S3DP3JIGXANCNFSM4HZH6WTQ>.

hartwork commented:
I thought about this some more and search the internet for `expat "XML_ATTR_INFO"` as well.
I conclude that no one uses that option publicly as of today.  Also the CI has never covered that option before.  It does not feel safe to me to just drop support for `XML_ATTR_INFO` as of today — someone might use it outside the public internet.

In my eyes to go forward we should:
- Keep `XML_ATTR_INFO` as a compile time switch for now, and keep it disabled by default
- Make the CI cover both `XML_ATTR_INFO` enabled and disabled

This happens to be in line with #244.

I'll review the latest changes on the base of this comment now.

hartwork commented on expat/CMakeLists.txt:
```diff
@@ -60,6 +60,7 @@ if(NOT WIN32)
 endif()
 option(XML_UNICODE "Use UTF-16 encoded chars (two bytes) instead of UTF-8" OFF)
 option(XML_UNICODE_WCHAR_T "Use wchar_t to represent UTF-16 instead of unsigned short" OFF)
+option(XML_ATTR_INFO "Define to allow retrieving the byte offsets for attribute names and values" ON)
```
Please turn default to `OFF`

MohammedKhajapasha replied:
done

hartwork commented on expat/configure.ac:
```diff
@@ -222,6 +222,17 @@ AC_DEFINE([XML_DTD], 1,
+AC_ARG_ENABLE([xml-attr-info],
+  [AS_HELP_STRING([--enable-xml-attr-info],
+                 [Enable retrieving the byte offsets for attribute names and values])], ,
+                 [enable_xml_attr_info=yes])
```
Please turn default to `no` and mention the default in the help text.

MohammedKhajapasha replied:
done, set XML_ATTR_INFO disable by default.

hartwork commented on expat/configure.ac:
```diff
@@ -222,6 +222,17 @@ AC_DEFINE([XML_DTD], 1,
+         [Define to allow retrieving the byte offsets for attribute names and values.])]
+      [test "x${enable_xml_attr_info}" = "xno"],
+      [AC_DEFINE([XML_ATTR_INFO], 0,
+         [Define to allow retrieving the byte offsets for attribute names and values.])])
```
Is there a way to resolve the code duplication regarding `AC_DEFINE([XML_ATTR_INFO]` and the error text here? `git grep AC_DEFINE` shows me it would be the first time we do/need this duplication.

MohammedKhajapasha replied:
will remove the false case i.e "no", as by default XML_ATTR_INFO is disable.

hartwork commented on expat/coverage.sh:
```diff
@@ -38,8 +38,11 @@ _get_coverage_dir() {
+       configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T -DXML_ATTR_INFO' )
+    else
+       configure_args+=( CPPFLAGS='-DXML_ATTR_INFO' )
+    fi
```
We need yet another switch for the CI — e.g. `attr_info_enabled` — here, to cover both enabled and disabled.

MohammedKhajapasha replied:
will change the passing xml-attr-info parameter to build & config so that the compilation take place for each case.

hartwork commented on expat/lib/libexpat.def:
```diff
@@ -72,7 +72,7 @@ EXPORTS
   XML_GetParsingStatus @65
 ; added with version 2.1.1
-; XML_GetAttributeInfo @66
+  XML_GetAttributeInfo @66
```
Please revert the uncommenting here — the default should remain at disabled.

MohammedKhajapasha replied:
Addressed review comments in latest patch set please do review.

MohammedKhajapasha commented:
Addressed below review comments in latest patchset.

--
Regards
Khajapasha
From: Sebastian Pipping [mailto:notifications@github.com]
Sent: Monday, July 1, 2019 4:09 AM
To: libexpat/libexpat <libexpat@noreply.github.com>
Cc: Khajapasha, Mohammed <mohammed.khajapasha@intel.com>; Author <author@noreply.github.com>
Subject: Re: [libexpat/libexpat] Enable XML_ATTR_INFO in build (#264)


@hartwork requested changes on this pull request.

Please check my comment at #264 (comment)<https://github.com/libexpat/libexpat/pull/264#issuecomment-507072890> and the review comments below. Thank you!

________________________________

In expat/CMakeLists.txt<https://github.com/libexpat/libexpat/pull/264#discussion_r298852097>:

> @@ -60,6 +60,7 @@ if(NOT WIN32)



 endif()



 option(XML_UNICODE "Use UTF-16 encoded chars (two bytes) instead of UTF-8" OFF)



 option(XML_UNICODE_WCHAR_T "Use wchar_t to represent UTF-16 instead of unsigned short" OFF)



+option(XML_ATTR_INFO "Define to allow retrieving the byte offsets for attribute names and values" ON)



Please turn default to OFF

________________________________

In expat/configure.ac<https://github.com/libexpat/libexpat/pull/264#discussion_r298852124>:

> @@ -222,6 +222,17 @@ AC_DEFINE([XML_DTD], 1,



 AC_DEFINE([XML_DEV_URANDOM], 1,



           [Define to include code reading entropy from `/dev/urandom'.])







+AC_ARG_ENABLE([xml-attr-info],



+  [AS_HELP_STRING([--enable-xml-attr-info],



+                 [Enable retrieving the byte offsets for attribute names and values])], ,



+                 [enable_xml_attr_info=yes])



Please turn default to no and mention the default in the help text.

________________________________

In expat/configure.ac<https://github.com/libexpat/libexpat/pull/264#discussion_r298852146>:

> @@ -222,6 +222,17 @@ AC_DEFINE([XML_DTD], 1,



 AC_DEFINE([XML_DEV_URANDOM], 1,



           [Define to include code reading entropy from `/dev/urandom'.])







+AC_ARG_ENABLE([xml-attr-info],



+  [AS_HELP_STRING([--enable-xml-attr-info],



+                 [Enable retrieving the byte offsets for attribute names and values])], ,



+                 [enable_xml_attr_info=yes])



+AS_IF([test "x${enable_xml_attr_info}" = "xyes"],



+      [AC_DEFINE([XML_ATTR_INFO], 1,



+         [Define to allow retrieving the byte offsets for attribute names and values.])]



+      [test "x${enable_xml_attr_info}" = "xno"],



+      [AC_DEFINE([XML_ATTR_INFO], 0,



+         [Define to allow retrieving the byte offsets for attribute names and values.])])



Is there a way to resolve the code duplication regarding AC_DEFINE([XML_ATTR_INFO] and the error text here? git grep AC_DEFINE shows me it would be the first time we do/need this duplication.

________________________________

In expat/coverage.sh<https://github.com/libexpat/libexpat/pull/264#discussion_r298852198>:

> @@ -38,8 +38,11 @@ _get_coverage_dir() {



 _configure() {



     local configure_args=()







-    ${unicode_enabled} \



-            && configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )



+    if [ "${unicode_enabled}" = true ]; then



+       configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T -DXML_ATTR_INFO' )



+    else



+       configure_args+=( CPPFLAGS='-DXML_ATTR_INFO' )



+    fi



We need yet another switch for the CI — e.g. attr_info_enabled — here, to cover both enabled and disabled.

________________________________

In expat/lib/libexpat.def<https://github.com/libexpat/libexpat/pull/264#discussion_r298852215>:

> @@ -72,7 +72,7 @@ EXPORTS



   XML_ResumeParser @64



   XML_GetParsingStatus @65



 ; added with version 2.1.1



-; XML_GetAttributeInfo @66



+  XML_GetAttributeInfo @66



Please revert the uncommenting here — the default should remain at disabled.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/libexpat/libexpat/pull/264?email_source=notifications&email_token=AMAHZCNZZXQFFBH6FTMXAVLP5EYZTA5CNFSM4HZH6WT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5B242A#pullrequestreview-256093800>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMAHZCI5IVWSPQ46JR5FSTTP5EYZTANCNFSM4HZH6WTQ>.

hartwork commented on expat/coverage.sh:
```diff
@@ -210,10 +214,11 @@ _main() {
+	        for xml_context in 0 1024 ; do
+	             _build_case
+		done
+	    done
```
Indentation looks off here, see `fi` on the line above. Also, please keep the empty line. Thanks!

MohammedKhajapasha replied:
Addressed review comments in latest patchset

hartwork commented on expat/coverage.sh:
```diff
@@ -41,6 +41,10 @@ _configure() {
             && configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )
 
+    if [ "${xml_attr_info}" = true ]; then
+       configure_args+=( --enable-xml-attr-info )
```
Indentation seems off here (3 versus 4 spaces).

MohammedKhajapasha replied:
Addressed review comments in latest patchset

hartwork commented on expat/coverage.sh:
```diff
@@ -41,6 +41,10 @@ _configure() {
     ${unicode_enabled} \
             && configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )
 
+    if [ "${xml_attr_info}" = true ]; then
```
Please use Bash `[[` (double square brackets) as the rest of `coverage.sh`. Thanks!

MohammedKhajapasha replied:
Addressed review comments in latest patchset

MohammedKhajapasha commented:
Hi,

I have updated #264 change with below review comments,
As per pull request #264, we have enabled XML_ATTR_INFO flag into auto make & cmake build system.
Can we have a separate issue for integrating XML_ATTR_INFO flag into CI, once above change has been merged ?
Created https://github.com/libexpat/libexpat/issues/273 issue to cover XML_ATTR_INFO into CI.

--
Regards
Khajapasha
From: Sebastian Pipping [mailto:notifications@github.com]
Sent: Tuesday, July 2, 2019 12:07 AM
To: libexpat/libexpat <libexpat@noreply.github.com>
Cc: Khajapasha, Mohammed <mohammed.khajapasha@intel.com>; Author <author@noreply.github.com>
Subject: Re: [libexpat/libexpat] Enable XML_ATTR_INFO in build (#264)


@hartwork requested changes on this pull request.

Looks pretty good to me by now, some formatting detail comments below.

Please not that to close #244<https://github.com/libexpat/libexpat/issues/244>, these new build flags will need to be used by the CI, which this pull request does not yet do. We can do it in here or in a new pull request after merging, as you like.

Best!

________________________________

In expat/coverage.sh<https://github.com/libexpat/libexpat/pull/264#discussion_r299166240>:

> @@ -210,10 +214,11 @@ _main() {

             if ${unicode_enabled} && ! ${with_mingw} ; then

                 continue

             fi

-

-            for xml_context in 0 1024 ; do

-                _build_case

-            done

+          for xml_attr_info in true false ; do

+              for xml_context in 0 1024 ; do

+                   _build_case

+             done

+          done

Indentation looks off here, see fi on the line above. Also, please keep the empty line. Thanks!

________________________________

In expat/coverage.sh<https://github.com/libexpat/libexpat/pull/264#discussion_r299166677>:

> @@ -41,6 +41,10 @@ _configure() {

     ${unicode_enabled} \

             && configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )



+    if [ "${xml_attr_info}" = true ]; then

+       configure_args+=( --enable-xml-attr-info )

Indentation seems off here (3 versus 4 spaces).

________________________________

In expat/coverage.sh<https://github.com/libexpat/libexpat/pull/264#discussion_r299167329>:

> @@ -41,6 +41,10 @@ _configure() {

     ${unicode_enabled} \

             && configure_args+=( CPPFLAGS='-DXML_UNICODE -DXML_UNICODE_WCHAR_T' )



+    if [ "${xml_attr_info}" = true ]; then

Please use Bash [[ (double square brackets) as the rest of coverage.sh. Thanks!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/libexpat/libexpat/pull/264?email_source=notifications&email_token=AMAHZCN3H7ZMJNZHG6LISULP5JFFVA5CNFSM4HZH6WT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5E5EBI#pullrequestreview-256496133>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMAHZCN6WYCS5UAYVUPGUQLP5JFFVANCNFSM4HZH6WTQ>.

hartwork commented:
Merged in df83b430f58e1f2c114cb3ee8915ba80234b073a with tiny changes on top, thanks!
